경북대 BE 김진현 - Step 2: 주문하기

kimjhyun0627 opened the pull request:
## 구현한 기능

- Step 1 2번째 리뷰를 반영해보았습니다.
- 영수증 역할을 할 `ORDERS` 테이블을 생성했습니다.
- 이에 맞는 도메인 및 VO 생성 & 도메인 테스트를 구현했습니다.
- 매 요청마다 `JwtHeader`에 `access token`을 함께 넘겨주는 것은 위험하기다고 생각해, `Member` 테이블에 새 필드를 추가하여 카카오 로그인 시 `access token`을 DB에 저장하여 사용할 수 있도록 변경했습니다.
- 기존 `JwtFilter`에서 `memberId` 값을 함께 넘겨줄 수 있도록 수정했습니다.
- 앞서 변경한 내용을 바탕으로 카카오톡 메세지 전송 클라이언트를 구현했습니다.
- 상품 주문 시, 위시리스트에 해당 상품이 존재한다면 삭제할 수 있게 해 보았습니다.

kimjhyun0627 committed:
docs: README 업데이트

kimjhyun0627 committed:
refactor: Step 1 코드리뷰 반영

- 카카오 로그인 POST -> GET
- @JsonNaming 반영으로 인한 개별 필드 어노테이션 삭제

kimjhyun0627 committed:
feat: 스키마 & 도메인 구현

- 테이블 스키마 구현
- 도메인 구현
- VO 구현

kimjhyun0627 committed:
feat: 레포지토리 구현

kimjhyun0627 committed:
feat: 클라이언트 구현 & Filter 수정

- 클라이언트 구현
- JwtFilter 및 JwtHeaderFilter 변경

kimjhyun0627 committed:
feat: 서비스 구현

- 서비스 구현
- 기존 서비스 일부 수정

kimjhyun0627 committed:
feat: 컨트롤러 & DTO 구현

- 컨트롤러 구현
- DTO 구현

kimjhyun0627 committed:
feat: 도메인 테스트 & 클라이언트 테스트 구현

kimjhyun0627 committed:
refactor: 불필요한 주석 삭제 & 메소드명 일부 변경

woodying commented on src/main/java/gift/controller/user/AuthController.java:
```diff
@@ -41,11 +42,10 @@ public ResponseEntity<AuthResponse> login(
-        AuthResponse response = memberService.kakaoLogin(authorizationCode.code());
+    @GetMapping("/login/kakao")
+    public ResponseEntity<AuthResponse> kakaoLogin(@RequestParam("code") String code) {
+        System.out.println("💬 Received kakao code: " + code);
```
테스트 후 지우시는걸 깜빡하신 것 같아요.
혹시 로그를 남기고 싶으신거라면 로거를 사용해주세요~! 🙂

추가로 테스트 중 값을 확인하고 싶다면 IDE 의 디버그 기능을 사용하는게 좋습니다. 
break point 를 찍고 code 값을 확인해볼 수 있어요.

kimjhyun0627 replied:
토큰 받아오는게 헷갈려서 테스트하다가 지우는걸 까먹었네요...;)
민감한 정보이니 로그로 남기기보다는 삭제해두겠습니다!

woodying commented on src/main/java/gift/entity/product/order/Order.java:
```diff
@@ -0,0 +1,76 @@
+    protected Order() {
+    }
+
+    private Order(ProductOption option, int quantity, String message) {
```
private 생성자 좋네요 👍

woodying commented on src/main/java/gift/external/kakao/message/KakaoMessageClient.java:
```diff
@@ -0,0 +1,69 @@
+                    String.class
+            );
+        } catch (Exception e) {
+            throw new RuntimeException("카카오 메시지 전송 실패", e);  // 💡 예외 래핑 추가
```
👍

woodying commented on src/main/java/gift/external/kakao/message/KakaoMessageClient.java:
```diff
@@ -0,0 +1,69 @@
+    private String generateSelfMessage(Order order, Member member) {
+        try {
+            Map<String, Object> message = new HashMap<>();
+            message.put("object_type", "text");
```
책임 분리를 위해 KakaoMessageClient를 감싸는 객체를 하나 만드는 것도 좋을 것 같아요 👍 
아니면 이 로직을 service에 두는 것도 좋을 것 같습니다 🙂

kimjhyun0627 replied:
책임 분리를 위해 `KakaoNotificationService`를 따로 생성해 기존 `KakaoMessegeClient`의 역할을 분리했습니다. `OrderService`에서는 `KakaoNotificationService`를 사용하도록 변경했습니다.

woodying commented on src/main/java/gift/external/kakao/token/dto/KakaoTokenResponse.java:
```diff
@@ -1,18 +1,14 @@
 import com.fasterxml.jackson.databind.PropertyNamingStrategies;
 import com.fasterxml.jackson.databind.annotation.JsonNaming;
 
 @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
```
👍

woodying reviewed (changes requested):
진현님 안녕하세요 🙂

요구사항 정말 빠르게 잘 구현해주셨네요! 👍👍

다른 부분들은 너무 잘 구현 해주셨는데, 한 가지 고쳐주셨으면 하는 부분이 있습니다.
`KakaoMessageClient` 와 관련된 부분인데요. 
이 객체는 **요청정보**를 받아서 **카카오 API 를 호출** 해주는 역할만 하도록 구현하는게 좋아보여요.  🙂

현재는 이 객체가 내부에서 order 와 member 객체를 이용해서 메시지를 생성하는 역할까지 맡고 있는 것으로 보입니다. 이 작업은 `KakaoMessageClient` 의 역할이 아닌 것 같아요~
책임 분리를 해주셨으면 좋겠습니다. 🙂

다른 부분은 너무 잘해주셨어요! 
구현하느라 고생 많으셨습니다. 👏
말씀드린 부분 수정해서 다시 리뷰 요청 부탁드려요~

kimjhyun0627 committed:
refactor: Step 2 코드리뷰 반영

- 불필요한 `sout` 삭제
- `KakaoMessageClient` 역할 분리: `KakaoNotificationService` 생성
- 테스트 코드 변경

woodying commented on src/main/java/gift/service/product/order/notification/KakaoNotificationService.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Map;
+import org.springframework.stereotype.Service;
+
+@Service
+public class KakaoNotificationService {
```
👍 👍

woodying commented on src/main/java/gift/external/kakao/message/KakaoMessageClient.java:
```diff
@@ -0,0 +1,40 @@
+        this.restTemplate = restTemplate;
+    }
+
+    public void sendToMe(String accessToken, String templateObjectJson) {
```
깔끔하게 토큰과 메시지만 전달받아서 메시지를 전송하는 역할만 수행하도록 변경되었네요! 👍

woodying reviewed (approved):
피드백 너무 잘 반영해주신 것 확인했습니다. 👍 👍 
앞으로도 코드 작성하실때 하나의 객체가 다양한 책임을 지지 않도록 더 신경써주시면 좋을 것 같아요! 🙂

2단계 구현 고생 많으셨습니다. 👏
